tests/gui/test_helper_find.py

```diff
@@ -0,0 +1,145 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the
+# Free Software Foundation, Inc.,
+# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+# ###################################################
+
+import unittest
+
+import mock
+
+from tests.gui.helper import GuiHelper
+
+
+class Widget(object):
+	"""
+	Fake widget with a name, children and possible a parent. All children of this widget will
+	get their parent attribute set to this widget.
+	"""
+	def __init__(self, name, children=None):
+		self.name = name
+		self.parent = None
+		self.children = children or []
+		for c in self.children:
+			c.parent = self
+
+	def __hash__(self):
+		return id(self)
+
+
+class FindWidgetTest(unittest.TestCase):
+	"""
+	Tests about finding widgets in pychan's internal lists.
+	"""
+	def setUp(self):
+		self.widgets = set()
+
+		def w(name, children=None):
+			widget = Widget(name, children)
+			self.widgets.add(widget)
+			return widget
+
+		self.w = w
+
+		helper = GuiHelper(mock.Mock(), mock.Mock())
+		self.find = lambda x: helper._find(self.widgets, x)
+
+	def test_simple(self):
+		"""
+		Single widget in the global list.
+		"""
+		target = self.w('target')
+		assert self.find('target') == target
+
+	def test_parent(self):
+		"""
+		Single widget with a child.
+		"""
+		target = self.w('target')
+		self.w('foobar', [
+			target
+		])
+
+		assert self.find('target') == target
+
+	def test_hierarchy(self):
+		"""
+		Target widget with multiple parents and another widget with the same name, but
+		without parent.
+		"""
+		target = self.w('target')
+
+		self.w('bla', [
+			self.w('baz', [
+				target
+			])
+		])
+		self.w('foobar')
+		self.w('target')
+
+		assert self.find('bla/baz/target') == target
+
+	def test_hierarchy_incomplete_path(self):
+		"""
+		Target with multiple parents, select it while omitting one parent in the search
+		path.
+		"""
+		target = self.w('target')
+
+		self.w('bla', [
+			self.w('baz', [
+				target
+			])
+		])
+		self.w('foobar')
+		self.w('target')
+
+		assert self.find('bla/target') == target
+
+	def test_hierarchy_not_absolute_path(self):
+		"""
+		Target with multiple parents, select it without specifing the top-most widget.
+		"""
+		target = self.w('target')
+
+		self.w('bla', [
+			self.w('baz', [
+				target
+			])
+		])
+		self.w('foobar')
+		self.w('target')
+
+		assert self.find('baz/target') == target
+
+	def test_wrong_search_path(self):
+		"""
+		A previous implementation had a bug where the code would go in a dead end when one
+		part of the path matches, but children below won't. Make sure this doesn't happen
+		again.
+		"""
+		target = self.w('target')
+
+		self.w('wrongpath', [
+			self.w('foobar')
+		])
+		self.w('foobar', [
+			target
+		])
+
+		assert self.find('foobar/target') == target
```

Comment:
I'm missing tests with multiple children on the same widget. Additionally, I'd like to see a test with a target widget having children, to see that processing children stops at the right time.